cmd/pinniped-proxy/src/pinniped.rs

```diff
@@ -239,88 +166,82 @@ async fn call_pinniped_exchange(
-    match token_creds
-        .create(&PostParams::default(), &cred_request)
-        .await
-    {
-        Ok(o) => Ok(o),
-        Err(e) => Err(anyhow::anyhow!(
-            "err creating token exchange: {:#?}\n{}",
-            serde_json::to_string(&cred_request).unwrap(),
-            e
-        )),
-    }
-}
+    // define the Api Resource dynamically
+    let gvk = GroupVersionKind::gvk(
+        &get_pinniped_login_api_group(),
+        TOKEN_REQUEST_VERSION,
+        TOKEN_REQUEST_KIND,
+    );
+    let ar = ApiResource::from_gvk(&gvk);
 
-/// call_pinniped_exchange returns the resulting TokenCredentialRequest with Status after requesting a token credential exchange.
-// TODO(agamez): remove this duplicated code and use a generic dynamic approach when possible.
-// see PR comments: https://github.com/kubeapps/kubeapps/pull/2515
-async fn call_pinniped_exchange_tmc(
-    authorization: &str,
-    k8s_api_server_url: &str,
-    k8s_api_ca_cert_data: &[u8],
-) -> Result<TokenCredentialRequestTMC> {
-    let pinniped_namespace = env::var(DEFAULT_PINNIPED_NAMESPACE)?;
-    let client = get_client_config(
-        k8s_api_server_url,
-        k8s_api_ca_cert_data,
-        pinniped_namespace.clone(),
-    )?;
-    let auth_token = match authorization.to_string().strip_prefix("Bearer ") {
-        Some(a) => a.to_string(),
-        None => authorization.to_string(),
-    };
-    let token_creds: Api<TokenCredentialRequestTMC> = Api::all(client.clone());
-    let cred_request = TokenCredentialRequestTMC::new(
-        "",
-        TokenCredentialRequestTMCSpec {
+    // create request
+    let cred_data = TokenCredentialRequest {
+        spec: TokenCredentialRequestSpec {
             token: Some(auth_token),
             authenticator: corev1::TypedLocalObjectReference {
-                name: env::var(DEFAULT_PINNIPED_AUTHENTICATOR_NAME).with_context(|| {
-                    format!("error retrieving {}", DEFAULT_PINNIPED_AUTHENTICATOR_NAME)
-                })?,
-                kind: env::var(DEFAULT_PINNIPED_AUTHENTICATOR_TYPE).with_context(|| {
-                    format!("error retrieving {}", DEFAULT_PINNIPED_AUTHENTICATOR_TYPE)
-                })?,
+                name: pinniped_auth_name.clone(),
+                kind: pinniped_auth_type.clone(),
                 api_group: Some(get_pinniped_authenticator_api_group().into()),
             },
         },
-    );
-
+        status: None,
+    };
+    let cred_request = DynamicObject::new("", &ar)
+        .within(&pinniped_namespace)
+        .data(serde_json::to_value(cred_data)?);
     debug!("{}", serde_json::to_string(&cred_request).unwrap());
-    match token_creds
+
+    // token credential request invocation
+    // we start first as a cluster-based call. if this call fails with a NotFound error, it is
+    // an indication we are on an old version which was namespace-based. we thus fallback to
+    // a namespace-based call.
```

Comment:
Great. I'd originally (incorrectly) though that we'd do this check once on startup rather than per request. But as you pointed out, it may differ between different clusters. I'd then thought that perhaps we could still do it on startup for all configured clusters and cache the results with a short life-time, but that won't be worthwhile either because, even though with the current solution, on TKG we'll have an extra request per proxied request, in reality we should be caching the credentials themselves so that we don't need a new token request per proxied request (separate #2875), which would mean that the extra request here would only be when we don't already have a valid cred for the cluster. Sorry, that's a  long-winded way of saying I think this is the best solution for now, thanks!